crates/tests/src/integration/masp.rs

```diff
@@ -2171,3 +2171,1068 @@ fn dynamic_assets() -> Result<()> {
+                NAM,
+                "--amount",
+                "1",
+                "--gas-limit",
+                "2000",
+                "--gas-price",
+                "1",
+                "--disposable-gas-payer",
+                "--ledger-address",
+                validator_one_rpc,
+            ],
+        )
+    });
+    assert!(captured.result.is_err());
+    _ = node.next_masp_epoch();
+    // sync shielded context
+    run(
+        &node,
+        Bin::Client,
+        vec!["shielded-sync", "--node", validator_one_rpc],
+    )?;
+    let captured = CapturedOutput::of(|| {
+        run(
+            &node,
+            Bin::Client,
+            vec![
+                "balance",
+                "--owner",
+                AA_VIEWING_KEY,
+                "--token",
+                NAM,
+                "--node",
+                validator_one_rpc,
+            ],
+        )
+    });
+    assert!(captured.result.is_ok());
+    assert!(captured.contains("nam: 50000"));
+
+    // 2. Valid masp fee payment
+    run(
+        &node,
+        Bin::Client,
+        vec![
+            "transfer",
+            "--source",
+            A_SPENDING_KEY,
+            "--target",
+            AB_PAYMENT_ADDRESS,
+            "--token",
+            NAM,
+            "--amount",
+            "10000",
+            "--gas-limit",
+            "20000",
+            "--gas-price",
+            "1",
+            "--disposable-gas-payer",
+            "--ledger-address",
+            validator_one_rpc,
+        ],
+    )?;
+    node.assert_success();
+    // sync shielded context
+    run(
+        &node,
+        Bin::Client,
+        vec!["shielded-sync", "--node", validator_one_rpc],
+    )?;
+    node.assert_success();
+    // Check the exact balance of the tx source to ensure that the masp fee
+    // payment transaction was executed only once
+    let captured = CapturedOutput::of(|| {
+        run(
+            &node,
+            Bin::Client,
+            vec![
+                "balance",
+                "--owner",
+                AA_VIEWING_KEY,
+                "--token",
+                NAM,
+                "--node",
+                validator_one_rpc,
+            ],
+        )
+    });
+    assert!(captured.result.is_ok());
+    assert!(captured.contains("nam: 20000"));
+    let captured = CapturedOutput::of(|| {
+        run(
+            &node,
+            Bin::Client,
+            vec![
+                "balance",
+                "--owner",
+                AB_VIEWING_KEY,
+                "--token",
+                NAM,
+                "--node",
+                validator_one_rpc,
+            ],
+        )
+    });
+    assert!(captured.result.is_ok());
+    assert!(captured.contains("nam: 10000"));
+
+    Ok(())
+}
+
+// Test that when paying gas via masp we select the gas limit as the minimum
+// between the transaction's gas limit and the protocol parameter.
+#[test]
+fn masp_fee_payment_gas_limit() -> Result<()> {
```

Comment:
this test doesn't seem to check that if gas limit provided by the tx is smaller than the protocol parameter, it uses the user supplied gas limit. Is this tested elsewhere?

Reply:
Ah it's actually already tested in `masp_fee_payment`